src/witness/yamlWitnessType.ml

```diff
@@ -0,0 +1,380 @@
+
+  let entry_type = "loop_invariant"
+
+  let to_yaml' {location; loop_invariant} =
+    [
+      ("location", Location.to_yaml location);
+      ("loop_invariant", Invariant.to_yaml loop_invariant);
+    ]
+
+  let of_yaml y =
+    let open GobYaml in
+    let* location = y |> find "location" >>= Location.of_yaml in
+    let+ loop_invariant = y |> find "loop_invariant" >>= Invariant.of_yaml in
+    {location; loop_invariant}
+end
+
+module PreconditionLoopInvariant =
+struct
+  type t = {
+    location: Location.t;
+    loop_invariant: Invariant.t;
+    precondition: Invariant.t;
+  }
+
+  let entry_type = "precondition_loop_invariant"
+
+  let to_yaml' {location; loop_invariant; precondition} =
+    [
+      ("location", Location.to_yaml location);
+      ("loop_invariant", Invariant.to_yaml loop_invariant);
+      ("precondition", Invariant.to_yaml precondition);
+    ]
+
+  let of_yaml y =
+    let open GobYaml in
+    let* location = y |> find "location" >>= Location.of_yaml in
+    let* loop_invariant = y |> find "loop_invariant" >>= Invariant.of_yaml in
+    let+ precondition = y |> find "precondition" >>= Invariant.of_yaml in
+    {location; loop_invariant; precondition}
+end
+
+module Target =
+struct
+  type t = {
+    uuid: string;
+    type_: string;
+    file_hash: string;
+  }
+
+  let to_yaml {uuid; type_; file_hash} =
+    `O [
+      ("uuid", `String uuid);
+      ("type", `String type_);
+      ("file_hash", `String file_hash);
+    ]
+
+  let of_yaml y =
+    let open GobYaml in
+    let* uuid = y |> find "uuid" >>= to_string in
+    let* type_ = y |> find "type" >>= to_string in
+    let+ file_hash = y |> find "file_hash" >>= to_string in
+    {uuid; type_; file_hash}
+end
+
+module Certification =
+struct
+  type t = {
+    string: string;
+    type_: string;
+    format: string;
+  }
+
+  let to_yaml {string; type_; format} =
+    `O [
+      ("string", `String string);
+      ("type", `String type_);
+      ("format", `String format);
+    ]
+
+  let of_yaml y =
+    let open GobYaml in
+    let* string = y |> find "string" >>= to_string in
+    let* type_ = y |> find "type" >>= to_string in
+    let+ format = y |> find "format" >>= to_string in
+    {string; type_; format}
+end
+
+module LoopInvariantCertificate =
+struct
+  type t = {
+    target: Target.t;
+    certification: Certification.t;
+  }
+
+  let entry_type = "loop_invariant_certificate"
+
+  let to_yaml' {target; certification} =
+    [
+      ("target", Target.to_yaml target);
+      ("certification", Certification.to_yaml certification);
+    ]
+
+  let of_yaml y =
+    let open GobYaml in
+    let* target = y |> find "target" >>= Target.of_yaml in
+    let+ certification = y |> find "certification" >>= Certification.of_yaml in
+    {target; certification}
+end
+
+module PreconditionLoopInvariantCertificate =
+struct
+  include LoopInvariantCertificate
+  let entry_type = "precondition_loop_invariant_certificate"
+end
+
+(* module type EntryType =
+sig
```

Comment:
Is it intentional to keep this commented out code?

Reply:
Not necessarily. It was an initial attempt to have less duplication and more type safety through a GADT, but for some reason this didn't work out (or at least as nicely).